File: apps/ios/Packages/FECore/Sources/FECore/DescriptionSanitizer.swift
import Foundation

/// Cleans event descriptions ingested from WordPress / Divi-built sources.
/// These often contain raw `[et_pb_*]` shortcodes and HTML markup that
/// users would otherwise see verbatim. Conservative: strips known noise
/// but preserves prose, line breaks, and links rendered as plain text.
public enum DescriptionSanitizer {
    /// Returns nil for nil input or whitespace-only output.
    public static func clean(_ raw: String?) -> String? {
        guard let raw, !raw.isEmpty else { return nil }
        var out = raw

        out = stripDiviShortcodes(out)
        out = stripGenericShortcodes(out)
        out = stripHTMLTags(out)
        out = decodeCommonEntities(out)
        out = collapseWhitespace(out)

        let trimmed = out.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    // MARK: - Pieces

    /// Match `[et_pb_*]` opening tags with arbitrary attributes and
    /// `[/et_pb_*]` closing tags. Divi shortcodes have unbounded attribute
    /// strings and may span multiple lines after rich text editing.
    static func stripDiviShortcodes(_ input: String) -> String {
        regexReplace(input, pattern: #"\[\/?et_pb_[a-z0-9_]*[^\]]*\]"#, with: "")
    }

    /// Generic WordPress shortcode fallback: `[caption …]`, `[gallery]`,
    /// `[/caption]`. Conservative — only matches `[name …]` where name is
    /// alphanumeric+underscore so we don't eat user-written `[notes]` prose.
    static func stripGenericShortcodes(_ input: String) -> String {
        regexReplace(input, pattern: #"\[\/?[a-z][a-z0-9_]*(?:\s[^\]]*)?\]"#, with: "")
    }

    /// Replaces `<br>` / `<br/>` / `<p>` with newlines, then strips all
    /// remaining tags. Preserves paragraph breaks in the visible text.
    static func stripHTMLTags(_ input: String) -> String {
        var s = input
        s = regexReplace(s, pattern: #"<br\s*\/?>"#, with: "\n", caseInsensitive: true)
        s = regexReplace(s, pattern: #"<\/p>"#, with: "\n\n", caseInsensitive: true)
        s = regexReplace(s, pattern: #"<[^>]+>"#, with: "")
        return s
    }

    static func decodeCommonEntities(_ input: String) -> String {
        var s = input
        let map: [(String, String)] = [
            ("&nbsp;", " "),
            ("&amp;", "&"),
            ("&lt;", "<"),
            ("&gt;", ">"),
            ("&quot;", "\""),
            ("&#039;", "'"),
            ("&apos;", "'"),
            ("&hellip;", "…"),
            ("&ndash;", "–"),
            ("&mdash;", "—"),
            ("&rsquo;", "'"),
            ("&lsquo;", "'"),
            ("&rdquo;", "\""),
            ("&ldquo;", "\""),
        ]
        for (entity, replacement) in map {
            s = s.replacingOccurrences(of: entity, with: replacement)
        }
        return s
    }

    /// Collapse runs of blank lines (3+) into a single blank line and
    /// trim trailing whitespace on each line.
    static func collapseWhitespace(_ input: String) -> String {
        var s = regexReplace(input, pattern: #"[ \t]+\n"#, with: "\n")
        s = regexReplace(s, pattern: #"\n{3,}"#, with: "\n\n")
        s = regexReplace(s, pattern: #"[ \t]{2,}"#, with: " ")
        return s
    }

    // MARK: - Helper

    private static func regexReplace(_ input: String, pattern: String, with replacement: String, caseInsensitive: Bool = true) -> String {
        var options: NSRegularExpression.Options = [.dotMatchesLineSeparators]
        if caseInsensitive { options.insert(.caseInsensitive) }
        guard let re = try? NSRegularExpression(pattern: pattern, options: options) else {
            return input
        }
        let range = NSRange(input.startIndex..<input.endIndex, in: input)
        return re.stringByReplacingMatches(in: input, options: [], range: range, withTemplate: replacement)
    }
}

File: packages/shared/src/description-sanitizer.ts
/**
 * Cleans event descriptions ingested from WordPress / Divi-built sources.
 * These often contain raw `[et_pb_*]` shortcodes and HTML markup that
 * users would otherwise see verbatim. Conservative — preserves prose,
 * strips known noise.
 *
 * Mirrors `supabase/functions/_shared/parsing.ts#cleanDescription` for
 * ingest-time cleanup, and the iOS `FECore.DescriptionSanitizer`. Web
 * runs this on read for descriptions written before the ingest fix
 * landed.
 */

const HTML_ENTITIES: ReadonlyArray<readonly [string, string]> = [
  ["&nbsp;", " "],
  ["&amp;", "&"],
  ["&lt;", "<"],
  ["&gt;", ">"],
  ["&quot;", '"'],
  ["&#039;", "'"],
  ["&apos;", "'"],
  ["&hellip;", "…"],
  ["&ndash;", "–"],
  ["&mdash;", "—"],
  ["&rsquo;", "'"],
  ["&lsquo;", "'"],
  ["&rdquo;", '"'],
  ["&ldquo;", '"'],
];

export function stripDiviShortcodes(value: string): string {
  return (
    value
      .replace(/\[\/?et_pb_[a-z0-9_]*[^\]]*\]/gis, "")
      // Handle the truncated trailing case: ingest used to slice raw
      // description to 500 chars before cleaning, sometimes severing a
      // shortcode mid-attribute. The unclosed `[et_pb_image src="..."` that
      // survives in those rows has no `]` so the rule above never matches.
      .replace(/\[\/?et_pb_[a-z0-9_]*[^\]]*$/is, "")
  );
}

/**
 * Generic WP shortcodes. Lowercase-only name so user prose like
 * `[See more details]` survives.
 */
export function stripGenericShortcodes(value: string): string {
  return (
    value
      .replace(/\[\/?[a-z][a-z0-9_]*(?:\s[^\]]*)?\]/gs, "")
      // Trailing unclosed shortcode (truncation case). Require a space
      // after the name so we don't eat user prose like `[See details` that
      // happens to be at end-of-string.
      .replace(/\[\/?[a-z][a-z0-9_]*\s[^\]]*$/s, "")
  );
}

export function stripHtmlPreservingBreaks(value: string): string {
  return value
    .replace(/<br\s*\/?>/gi, "\n")
    .replace(/<\/p>/gi, "\n\n")
    .replace(/<[^>]+>/g, "");
}

export function decodeHtmlEntities(value: string): string {
  let out = value;
  for (const [entity, replacement] of HTML_ENTITIES) {
    out = out.split(entity).join(replacement);
  }
  return out;
}

export function collapseWhitespace(value: string): string {
  return value
    .replace(/[ \t]+\n/g, "\n")
    .replace(/\n{3,}/g, "\n\n")
    .replace(/[ \t]{2,}/g, " ");
}

/**
 * Returns null when the cleaned result is empty / whitespace-only,
 * so callers can hide the "About" section entirely.
 */
export function cleanDescription(raw: string | null | undefined): string | null {
  if (!raw) {
    return null;
  }
  let out = raw;
  out = stripDiviShortcodes(out);
  out = stripGenericShortcodes(out);
  out = stripHtmlPreservingBreaks(out);
  out = decodeHtmlEntities(out);
  out = collapseWhitespace(out);
  const trimmed = out.trim();
  return trimmed.length === 0 ? null : trimmed;
}
